### PR TITLE
Defactor hotplug

### PR DESCRIPTION
**Pull request by itdaniher**

### CODE REVIEW

File: SMU.cpp [hunk 1]
```diff
@@ -90,30 +92,57 @@ void SessionItem::start()
...
+  qDebug() << "detached\n";
   qDebug() << m_devices;
-  for (auto i: m_session->m_available_devices) {
-    auto dev = m_session->add_device(&*i);
-    m_devices.append(new DeviceItem(this, dev));
+  if (!m_active) {
+      m_session->remove_device(device);
+  }
+  else {
+      this->cancel();
```
Comment (kevinmehall) [inline SMU.cpp]: need to `wait` for the transfers to actually cancel. Also, don't you still need to then call remove_device?

File: SMU.cpp [hunk 2]
```diff
@@ -90,30 +92,57 @@ void SessionItem::start()
...
+  }
+  if (m_session->m_devices.size() < m_devices.size()) {
+    QList<DeviceItem *> devices;
+    m_devices.swap(devices);
```
Comment (kevinmehall) [inline SMU.cpp]: Shouldn't this remove just the relevant device instead of clearing all devices?

File: SMU.cpp [hunk 3]
```diff
@@ -90,30 +92,57 @@ void SessionItem::start()
...
+    devicesChanged();
```
Comment (kevinmehall) [inline SMU.cpp]: Why emit devicesChanged twice?

**Comment by itdaniher** [comment]
@kevinmehall fixes pushed

**Comment by itdaniher** [comment]
continuous mode isn't finished yet, but does work on windows with the most recent build

**Comment by itdaniher** [comment]
@kevinmehall additional thoughts?

**Pull request merged**